RemoteGraphicsContextGLProxy::initializeIPC makes a copy of the StreamServerConnection::Handle

mattwoodrow opened the pull request:
#### 50f865ffbe32f1fc8a092583f719630b62ae7604
<pre>
RemoteGraphicsContextGLProxy::initializeIPC makes a copy of the StreamServerConnection::Handle
<a href="https://bugs.webkit.org/show_bug.cgi?id=257519">https://bugs.webkit.org/show_bug.cgi?id=257519</a>
&lt;rdar://109720377&gt;

Reviewed by Kimmo Kinnunen.

The GPUConnectionToWebProcess::CreateGraphicsContextGL object&apos;s constructor doesn&apos;t actually move, so we end up making a copy for the IPC message.

The &apos;serverConnectionHandle&apos; instance is then still alive, and holding on to a MachSendRight.

waitUntilInitialized can then block (for the 30 second timeout) if the GPUP crashes, since notification of the crash doesn&apos;t happen until all send rights are destroyed (including the one on the stack).

This change ensures that our copy of the handle is cleared, before trying to wait on the connection.

* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.cpp:
(WebKit::RemoteGraphicsContextGLProxy::initializeIPC):

Canonical link: <a href="https://commits.webkit.org/264766@main">https://commits.webkit.org/264766@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a0f19907204b175543389770488ef686af240a36

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/8547 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/8837 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/9051 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/10203 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/8574 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/10826 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/8802 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/11440 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/8693 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/9721 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/7686 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/10358 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/7021 "Passed tests") | | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/15356 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/8140 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/7965 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/11310 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/8442 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/6901 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/7719 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2077 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/11928 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/8184 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kkinnunen-apple commented on Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.cpp:
```diff
@@ -111,7 +111,10 @@ RemoteGraphicsContextGLProxy::RemoteGraphicsContextGLProxy(IPC::Connection& conn
+        IPC::StreamServerConnection::Handle handle = WTFMove(serverConnectionHandle);
+        m_connection->send(Messages::GPUConnectionToWebProcess::CreateGraphicsContextGL(contextAttributes(), m_graphicsContextGLIdentifier, renderingBackend.ensureBackendCreated(), WTFMove(handle)), 0, IPC::SendOption::DispatchMessageEvenWhenWaitingForSyncReply);
+    }
     m_streamConnection->open(*this, renderingBackend.dispatcher());
```
alternatively you can do
```
serverConnectionHandle = { } ; // FIXME(https://bugs.webkit.org/show_bug.cgi?id=257528): Support real rvalue semantics for send 
```

kkinnunen-apple reviewed (approved):
Uh-oh. I thought it was benign to not support it, but apparently not!
Should implement it.

mattwoodrow committed:
RemoteGraphicsContextGLProxy::initializeIPC makes a copy of the StreamServerConnection::Handle
https://bugs.webkit.org/show_bug.cgi?id=257519
<rdar://109720377>

Reviewed by Kimmo Kinnunen.

The GPUConnectionToWebProcess::CreateGraphicsContextGL object's constructor doesn't actually move, so we end up making a copy for the IPC message.

The 'serverConnectionHandle' instance is then still alive, and holding on to a MachSendRight.

waitUntilInitialized can then block (for the 30 second timeout) if the GPUP crashes, since notification of the crash doesn't happen until all send rights are destroyed (including the one on the stack).

This change ensures that our copy of the handle is cleared, before trying to wait on the connection.

* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.cpp:
(WebKit::RemoteGraphicsContextGLProxy::initializeIPC):

Canonical link: https://commits.webkit.org/264766@main
